### PR TITLE
Allow configuration of HTTP timeouts for maven repositories

### PR DESCRIPTION
**Pull request by ammachado**
## What's changed?
Changed maven downloader to allow configuration of connection/read timeouts

## What's your motivation?
https://maven.apache.org/guides/mini/guide-http-settings.html#connection-timeouts

## Anything in particular you'd like reviewers to focus on?
No

## Anyone you would like to review specifically?
No

## Have you considered any alternatives or workarounds?
No

## Any additional context
N/A

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: rewrite-maven/src/main/java/org/openrewrite/maven/tree/MavenRepository.java [hunk 1]
```diff
@@ -79,14 +80,51 @@ public class MavenRepository implements Serializable {
...
+     * @deprecated Use {@link #MavenRepository(String, String, String, String, boolean, String, String, Duration, Duration, Boolean)}
+     */
+    @Deprecated
     @JsonIgnore
     public MavenRepository(
             @Nullable String id, String uri, @Nullable String releases, @Nullable String snapshots, boolean knownToExist,
             @Nullable String username, @Nullable String password, @Nullable Boolean deriveMetadataIfMissing
+    ) {
+        this(id, uri, releases, snapshots, knownToExist, username, password, null, null, deriveMetadataIfMissing);
+    }
```
Comment (ammachado) [inline rewrite-maven/src/main/java/org/openrewrite/maven/tree/MavenRepository.java:121]: This is used by an external project, `org.openrwrite.gradle.tooling:model`.

**Comment by timtebeek** [comment]
Nice addition @ammachado ! I'm tagging @sambsnyd for an additional review as we're adding fields to tree classes, and he reworked parts of the Maven repository handling recently.

Also copying the details on the failed test to save others a click
```
MavenPomDownloaderTest > centralIdOverridesDefaultRepository() FAILED
    java.lang.AssertionError: 
    Expecting elements:
      [MavenRepository(id=local, uri=file:///home/runner/.m2/repository, releases=true, snapshots=true, knownToExist=true, username=null, *** connectTimeout=null, readTimeout=null, deriveMetadataIfMissing=false),
        MavenRepository(id=central, uri=https://repo.maven.apache.org/maven2, releases=true, snapshots=false, knownToExist=true, username=null, *** connectTimeout=null, readTimeout=null, deriveMetadataIfMissing=true)]
    to be exactly 1 times URI https://internalartifactrepository.yourorg.com/
        at org.openrewrite.maven.internal.MavenPomDownloaderTest.centralIdOverridesDefaultRepository(MavenPomDownloaderTest.java:115)
```

**Comment by ammachado** [comment]
> Nice addition @ammachado ! I'm tagging @sambsnyd for an additional review as we're adding fields to tree classes, and he reworked parts of the Maven repository handling recently.
> 
> Also copying the details on the failed test to save others a click
> 
> ```
> MavenPomDownloaderTest > centralIdOverridesDefaultRepository() FAILED
>     java.lang.AssertionError: 
>     Expecting elements:
>       [MavenRepository(id=local, uri=file:///home/runner/.m2/repository, releases=true, snapshots=true, knownToExist=true, username=null, *** connectTimeout=null, readTimeout=null, deriveMetadataIfMissing=false),
>         MavenRepository(id=central, uri=https://repo.maven.apache.org/maven2, releases=true, snapshots=false, knownToExist=true, username=null, *** connectTimeout=null, readTimeout=null, deriveMetadataIfMissing=true)]
>     to be exactly 1 times URI https://internalartifactrepository.yourorg.com/
>         at org.openrewrite.maven.internal.MavenPomDownloaderTest.centralIdOverridesDefaultRepository(MavenPomDownloaderTest.java:115)
> ```

As a suggestion, can you consider publishing the test reports on a failed build? Perhaps using a custom action for it (for example, https://github.com/marketplace/actions/junit-report-action).

**Comment by timtebeek** [comment]
> As a suggestion, can you consider publishing the test reports on a failed build? Perhaps using a custom action for it (for example, https://github.com/marketplace/actions/junit-report-action).

Yes I'd been eyeing https://github.com/gradle/actions/tree/main/setup-gradle#build-reporting earlier today, as part of our upgrade in https://github.com/openrewrite/gh-automation/ . It's been a pain point of mine as well to have to go into the logs.

**Comment by sambsnyd** [comment]
What's the motivation here? Does this solve a problem you are running into?

**Comment by cjohnsto88** [comment]
Not my PR but I am watching as I also have an issue that would be solved with this.

During execution the plugin downloads `maven-metadata.xml` files. My company's internal proxy of Maven Central is very slow for downloads of these files. I cannot get the wider and obvious root cause fixed. The files do download _eventually_, however in the current implementation there is no way to set the timeout as the custom code to download the `maven-metadata.xml` files in the plugin does not respect the `<configuration>` declarations for a `<server>` specified [here](https://maven.apache.org/guides/mini/guide-http-settings.html#connection-timeouts).

**Comment by ammachado** [comment]
> Not my PR but I am watching as I also have an issue that would be solved with this.
> 
> During execution the plugin downloads `maven-metadata.xml` files. My company's internal proxy of Maven Central is very slow for downloads of these files. I cannot get the wider and obvious root cause fixed. The files do download _eventually_, however in the current implementation there is no way to set the timeout as the custom code to download the `maven-metadata.xml` files in the plugin does not respect the `<configuration>` declarations for a `<server>` specified [here](https://maven.apache.org/guides/mini/guide-http-settings.html#connection-timeouts).

That's exactly the issue that originated this change.

**Comment by timtebeek** [comment]
Thanks a lot @ammachado ! Should help to conform to the user's configured session, as we've seen more variation in artifact repository performance than we had expected folks would tolerate in their day to day work. 🙃

File: rewrite-maven/src/test/java/org/openrewrite/maven/MavenSettingsTest.java [hunk 2]
```diff
@@ -334,13 +340,41 @@ void serverCredentials() {
...
+                <settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
+                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                    xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0 http://maven.apache.org/xsd/settings-1.0.0.xsd">
+                      <servers>
+                        <server>
+                          <id>server001</id>
+                          <configuration>
+                            <connectTimeout>40000</connectTimeout>
+                            <readTimeout>50000</readTimeout>
+                          </configuration>
```
Comment (timtebeek) [inline rewrite-maven/src/test/java/org/openrewrite/maven/MavenSettingsTest.java:357]: I'm having another look at https://maven.apache.org/guides/mini/guide-http-settings.html#connection-timeouts
Over there it seems there's two options
## A. an older `<timeout>` directly in `<configuration>`
```xml
<settings>
  <servers>
    <server>
      <id>my-server</id>
      <configuration>
        <timeout>6000</timeout> <!-- milliseconds -->
      </configuration>
    </server>
  </servers>
</settings>
```
## B. a newer `<connectionTimeout>` and `<readTimeout>` nested inside `<httpConfiguration><put>`
```xml
<settings>
  <servers>
    <server>
      <id>my-server</id>
      <configuration>
        <httpConfiguration>
          <put>
            <connectionTimeout>10000</connectionTimeout> <!-- milliseconds -->
          </put>
        </httpConfiguration>
      </configuration>
    </server>
  </servers>
</settings>
```

## In this PR
We seem to have added `connectTimeout` (as opposed to `connectionTimeout`, as well as `readTimeout`, but neither of those nested in the `<httpConfiguration>` + `<put>` elements I'd expect from the Maven docs.

In practice I think these fields won't be set when reading a Maven settings file from disk in style A or B above. Should we make any adjustments here @ammachado ?

Comment (timtebeek) [reply rewrite-maven/src/test/java/org/openrewrite/maven/MavenSettingsTest.java:357]: Picked up in 
- https://github.com/openrewrite/rewrite/pull/4302

**Pull request merged**